pallets/oracle-collection/src/lib.rs

```diff
@@ -385,67 +351,67 @@ pub mod pallet {
 		}
 
 		fn register_id(key: &T::OracleKey, collection_id: &T::CollectionId) -> DispatchResult {
-			Keys::<T>::mutate(collection_id, key, |info| {
-				if info.usage_refs.is_zero() {
+			Keys::<T>::mutate(collection_id, key, |usage_refs| {
+				if usage_refs.is_zero() {
 					Self::deposit_event(Event::<T>::AddedKey {
 						collection_id: *collection_id,
 						key: *key,
 					});
+
+					CollectionKeyCount::<T>::mutate(collection_id, |counter| -> DispatchResult {
+						if *counter == T::MaxCollectionSize::get() {
+							Err(Error::<T>::MaxCollectionSize)?;
+						}
+
+						counter.ensure_add_assign(1)?;
+
+						Ok(())
+					})?;
 				}
 
-				info.usage_refs.ensure_add_assign(1)?;
+				usage_refs.ensure_add_assign(1)?;
 				Ok(())
 			})
 		}
 
 		fn unregister_id(key: &T::OracleKey, collection_id: &T::CollectionId) -> DispatchResult {
-			Self::mutate_and_remove_keys_if_clean(*collection_id, *key, |info| {
-				info.usage_refs
+			Keys::<T>::mutate_exists(collection_id, key, |maybe_info| {
+				let usage_refs = maybe_info.get_or_insert(Default::default());
+
+				usage_refs
 					.ensure_sub_assign(1)
 					.map_err(|_| Error::<T>::KeyNotRegistered)?;
 
-				if info.usage_refs.is_zero() {
+				if usage_refs.is_zero() {
 					Self::deposit_event(Event::<T>::RemovedKey {
 						collection_id: *collection_id,
 						key: *key,
 					});
-				}
 
-				Ok(())
-			})
-		}
-	}
+					CollectionKeyCount::<T>::mutate_exists(
+						collection_id,
+						|maybe_counter| -> DispatchResult {
+							let counter = maybe_counter.get_or_insert(Default::default());
 
-	impl<T: Config> Pallet<T> {
-		fn mutate_and_remove_keys_if_clean(
-			collection_id: T::CollectionId,
-			key: T::OracleKey,
-			f: impl FnOnce(&mut KeyInfo<T>) -> DispatchResult,
-		) -> DispatchResult {
-			Keys::<T>::mutate_exists(collection_id, key, |maybe_info| {
-				let info = maybe_info.get_or_insert(Default::default());
+							counter.ensure_sub_assign(1)?;
+
+							if counter.is_zero() {
+								*maybe_counter = None;
+							}
```

Comment:
Is this correct? Isn't usage refs, unequal `usafe_refs`?

Reply:
I do not understand the question; could you extend it?

To recap:
- `usage_refs` corresponds to how many times a single key is used.
- `counter` corresponds to how many keys are used in the collection

Reply:
Ahh, I thought you are removing the key, but you are removing the counter. Get it.